gadgets/traceloop/go/program.go

```diff
@@ -0,0 +1,835 @@
+	Comm               [16]uint8
+	ContNr             uint8
+	_                  [62]byte
+}
+
+type syscallEvent struct {
+	bootTimestamp      uint64
+	monotonicTimestamp uint64
+	typ                eventType
+	contNr             uint8
+	cpu                uint16
+	id                 uint16
+	pid                uint32
+	comm               string
+	args               []uint64
+	mountNsID          uint64
+	retval             uint64
+}
+
+type syscallEventContinued struct {
+	monotonicTimestamp uint64
+	index              uint8
+	param              string
+}
+
+type syscallParam struct {
+	name    string
+	value   string
+	content *string
+}
+
+type event struct {
+	timestamp  int64
+	mountNsID  uint64
+	cpu        uint16
+	pid        uint32
+	comm       string
+	syscall    string
+	parameters []syscallParam
+	retval     string
+}
+
+func paramsToString(parameters []syscallParam) string {
+	var sb strings.Builder
+
+	for idx, p := range parameters {
+		value := p.value
+		if p.content != nil {
+			value = *p.content
+		}
+
+		sb.WriteString(p.name)
+		sb.WriteRune('=')
+		sb.WriteString(value)
+
+		if idx < len(parameters)-1 {
+			sb.WriteString(", ")
+		}
+	}
+
+	return sb.String()
+}
+
+func (t *tracelooper) attach(mntnsID uint64) error {
+	perfBufferName := fmt.Sprintf("perf_buffer_%d", mntnsID)
+
+	// 1. Create inner Map as perf buffer.
+	// Keep the spec in sync with program.bpf.c.
+	innerBuffer, err := api.NewMap(api.MapSpec{
+		Name:      perfBufferName,
+		Type:      api.PerfEventArray,
+		KeySize:   uint32(4),
+		ValueSize: uint32(4),
+	})
+	if err != nil {
+		return fmt.Errorf("creating map %s", fmt.Sprintf("perf_buffer_%d", mntnsID))
+	}
+
+	// 2. Use this inner Map to create the perf reader.
+	perfReader, err := api.NewPerfReader(innerBuffer, 64*linuxPageSize, true)
+	if err != nil {
+		innerBuffer.Close()
+
+		return fmt.Errorf("creating perf ring buffer: %w", err)
+	}
+
+	// 3. Add the inner map's file descriptor to outer map.
+	err = t.mapOfPerfBuffers.Update(mntnsID, innerBuffer, api.UpdateNoExist)
+	if err != nil {
+		innerBuffer.Close()
+		perfReader.Close()
+
+		return fmt.Errorf("adding perf buffer to map with mntnsID %d: %w", mntnsID, err)
+	}
+
+	t.readers.Store(mntnsID, &containerRingReader{
+		innerBuffer: innerBuffer,
+		perfReader:  perfReader,
+	})
+
+	return nil
+}
+
+func (t *tracelooper) detach(mntnsID uint64) error {
+	err := t.mapOfPerfBuffers.Delete(mntnsID)
+	if err != nil {
+		return fmt.Errorf("removing perf buffer from map with mntnsID %d: %w", mntnsID, err)
+	}
+
+	return nil
+}
+
+func fromCString(in []byte) string {
+	for i := 0; i < len(in); i++ {
+		if in[i] == 0 {
+			return string(in[:i])
+		}
+	}
+	return string(in)
+}
+
+func fromCStringN(in []byte, length int) string {
+	l := len(in)
+	if length < l {
+		l = length
+	}
+
+	for i := 0; i < l; i++ {
+		if in[i] == 0 {
+			return string(in[:i])
+		}
+	}
+	return string(in[:l])
+}
+
+func wallTimeFromBootTime(ts uint64) int64 {
+	if ts == 0 {
+		return 0
+	}
```

Comment:
Could be deleted since the line `return time.Unix...` will do the same thing.
Then, wallTimeFromBootTime could be merged in timestampFromEvent